### ISSUE TITLE
Investigate changes in image size after switching to applying modules individually

### ISSUE THREAD

**Issue by goldmann** [cekit/cekit#371]
This is related to #329.

We need to ensure that the resulting image is not bigger than the original. If it is, we need to investigate what is the root cause and what can we do about it.

**Comment by goldmann** [comment]
Initial results of OpenJDK image build in OSBS:

After #329? | Squashed? | ppc64le | x86_64
------------ | ------------- | ------------- | -------------
Before | No | 667.89 MB | 658.74 MB
Before | Yes | 480.26 MB | 452.30 MB
After | No | 1464.27 MB | 1793.87 MB
After | Yes | 884.92 MB | 992.29 MB

It can be clearly seen that the image size was **doubled**. This needs immediate action and review. This is a blocker for the release.

/cc @rcernich @jmtd

**Comment by goldmann** [comment]
Modules merged before tar:

```
   Date      Time    Attr         Size   Compressed  Name
------------------- ----- ------------ ------------  ------------------------
2019-01-23 15:30:19 .....         7440         7680  069f365bd866ea7cb1482e59b62a42456cd6fff0de4846516307e7f6bc6b58cc.json
2019-01-23 15:30:19 D....            0            0  27080e9653ee4f09c4479c88303b25db5893727f4e06b4b90663957b07284ec5
2019-01-23 15:30:19 .....            3          512  27080e9653ee4f09c4479c88303b25db5893727f4e06b4b90663957b07284ec5/VERSION
2019-01-23 15:30:19 .....         7144         7168  27080e9653ee4f09c4479c88303b25db5893727f4e06b4b90663957b07284ec5/json
2019-01-23 15:30:19 .....    260648960    260648960  27080e9653ee4f09c4479c88303b25db5893727f4e06b4b90663957b07284ec5/layer.tar
2019-01-23 15:30:19 D....            0            0  9b66bc9c64bae25ae7919c518c58fee6f254d9833358f18510586981ac49e40b
2019-01-23 15:30:19 .....            3          512  9b66bc9c64bae25ae7919c518c58fee6f254d9833358f18510586981ac49e40b/VERSION
2019-01-23 15:30:19 .....          394          512  9b66bc9c64bae25ae7919c518c58fee6f254d9833358f18510586981ac49e40b/json
2019-01-23 15:30:19 .....    213596160    213596160  9b66bc9c64bae25ae7919c518c58fee6f254d9833358f18510586981ac49e40b/layer.tar
2019-01-23 15:30:19 D....            0            0  d779d541f00beb04e5362327ba8007c5a19c65ddbf69c2e0ead54649687bc728
2019-01-23 15:30:19 .....            3          512  d779d541f00beb04e5362327ba8007c5a19c65ddbf69c2e0ead54649687bc728/VERSION
2019-01-23 15:30:19 .....          470          512  d779d541f00beb04e5362327ba8007c5a19c65ddbf69c2e0ead54649687bc728/json
2019-01-23 15:30:19 .....        10240        10240  d779d541f00beb04e5362327ba8007c5a19c65ddbf69c2e0ead54649687bc728/layer.tar
                    .....          426          512  manifest.json
                    .....          159          512  repositories
------------------- ----- ------------ ------------  ------------------------
2019-01-23 15:30:19          474271402    474273792  12 files, 3 folders
```

Modules applied individually tar:

```
   Date      Time    Attr         Size   Compressed  Name
------------------- ----- ------------ ------------  ------------------------
2019-01-25 09:21:17 D....            0            0  42f4a29c8fcd53e182513dba6e25054304646e5430ba36759c52a7a6b33c0891
2019-01-25 09:21:17 .....            3          512  42f4a29c8fcd53e182513dba6e25054304646e5430ba36759c52a7a6b33c0891/VERSION
2019-01-25 09:21:17 .....         7267         7680  42f4a29c8fcd53e182513dba6e25054304646e5430ba36759c52a7a6b33c0891/json
2019-01-25 09:21:17 .....    826869760    826869760  42f4a29c8fcd53e182513dba6e25054304646e5430ba36759c52a7a6b33c0891/layer.tar
2019-01-25 09:21:17 .....         7564         7680  586cb1cd25ff7f291ae11fcce88bfc6afbd9c7afb0bf4197b046fd2932297901.json
2019-01-25 09:21:17 D....            0            0  7173f91cfb7af606b8ed53b0326605310e83547b621ca3953f52d66ed2b44c7d
2019-01-25 09:21:17 .....            3          512  7173f91cfb7af606b8ed53b0326605310e83547b621ca3953f52d66ed2b44c7d/VERSION
2019-01-25 09:21:17 .....          471          512  7173f91cfb7af606b8ed53b0326605310e83547b621ca3953f52d66ed2b44c7d/json
2019-01-25 09:21:17 .....        10240        10240  7173f91cfb7af606b8ed53b0326605310e83547b621ca3953f52d66ed2b44c7d/layer.tar
2019-01-25 09:21:17 D....            0            0  917a3ccba0d541fff5b4ee3b539d480d325752f2fa9a21bb8cbba6d39f7c0cbd
2019-01-25 09:21:17 .....            3          512  917a3ccba0d541fff5b4ee3b539d480d325752f2fa9a21bb8cbba6d39f7c0cbd/VERSION
2019-01-25 09:21:17 .....          395          512  917a3ccba0d541fff5b4ee3b539d480d325752f2fa9a21bb8cbba6d39f7c0cbd/json
2019-01-25 09:21:17 .....    213596160    213596160  917a3ccba0d541fff5b4ee3b539d480d325752f2fa9a21bb8cbba6d39f7c0cbd/layer.tar
                    .....          479          512  manifest.json
                    .....          212          512  repositories
------------------- ----- ------------ ------------  ------------------------
2019-01-25 09:21:17         1040492557   1040495104  12 files, 3 folders
```

**Comment by goldmann** [comment]
We can see here 3 layer files, first two are from the RHEL 7 base image, the third one is the squashed OpenJDK image. This is where we can see over 3 time bump in the size (3.17 to be exact).

**Comment by goldmann** [comment]
Here are lists of files included in the interesting layers.

* [27080e9653ee4f09c4479c88303b25db5893727f4e06b4b90663957b07284ec5/layer.tar](https://github.com/cekit/cekit/files/2795412/before-files.txt)
* [42f4a29c8fcd53e182513dba6e25054304646e5430ba36759c52a7a6b33c0891/layer.tar](https://github.com/cekit/cekit/files/2795413/after-files.txt)

**Comment by goldmann** [comment]
We can see that after applying modules individually number of files grew by 337 from 17111 to 17448.

New build contains 3 more packages:

* java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64
* java-1.8.0-openjdk-devel-1.8.0.191.b12-1.el7_6.x86_64
* java-1.8.0-openjdk-headless-1.8.0.191.b12-1.el7_6.x86_64

⚠️ **This may be pulled because of some improper requirements**. Needs further investigation!

A difference for these files can be found here: https://www.diffchecker.com/LzTFiPr9.

### Notes

* There are changes in the alternatives. Some files are removed, but many files are added. Related probably to installing two JDKs.
* Together with alternatives, we are missing or adding binaries. Related probably to installing two JDKs.
* The whole directory `usr/lib/jvm-exports/java-1.8.0` and similar (`usr/lib/jvm/java-1.8.0`, `usr/lib/jvm/jre-1.8.0`), related to JDK 8.
* There are some unexpected `var/cache/yum/x86_64/7Server` files as well as other YUM/RPM related files.

### Summary

Besides the unexpected JDK 8 in there, there are actually no obvious other mistakes. The `/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64` directory weights 166 MB.

Additional `/var/lib/yum/rpmdb-indexes/`, `/var/lib/yum/history/` directories are very small.

**Comment by goldmann** [comment]
Difference in packages can be found here: https://www.diffchecker.com/z4oZyBNA.

**Comment by goldmann** [comment]
It looks that the JDK8 was installed previously too, but later, in scripts, it was removed by running this script: https://github.com/jboss-openshift/cct_module/blob/sprint-28/jboss/container/openjdk/jdk/11/configure.sh.

@jmtd The issue is in ordering - to be able to remove JDK 8, when 11 should be the only one installed, the removal script needs to be run **after** Maven install, which pulls the JDK 8. Where should I report it?

**Comment by goldmann** [comment]
This will explain only part of the issue, the other part is in the `/var/cache/yum/x86_64/7Server/rhel-7-server-rpms/gen/primary_db.sqlite` which weights **283 MB**! The whole `/var/cache/yum` weights **375 MB**... It looks like `yum clean all` was not executed and this is actually a bug. Created #372.

**Comment by goldmann** [comment]
With the fix proposed in #373 I kicked a new build of the image to see how big it will be. I expect it to be around 650 MB for ppc64le and 620 MB for x86_64. This should go even further down when the module issue described in the comment above (https://github.com/cekit/cekit/issues/371#issuecomment-457534452) will be fixed.

**Comment by rcernich** [comment]
@goldmann, there is a PR that needs to be merged in cct_module, jboss-openshift/cct_module#322

**Comment by rcernich** [comment]
FWIW, I tested this with both the Java S2I and EAP images and there was a negligible difference in size.

**Comment by goldmann** [comment]
@rcernich Thanks, I forgot about this https://github.com/jboss-openshift/cct_module/pull/322 PR!

With my fix in PR #373, sizes are following: ppc64le: 625.23 MB and 617.63 MB for the x86_64 version, which is even less what I predicted above, good. When the PR to cct_module repository will be merged, everything should be OK.

Next step is to try it with a different image and compare results.
